sbom/spdx/encoder.go

```diff
@@ -0,0 +1,299 @@
+package spdx
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	spdxjson "github.com/spdx/tools-golang/json"
+	"io"
+	"time"
+
+	"github.com/quay/claircore"
+	"github.com/quay/claircore/sbom"
+
+	"github.com/spdx/tools-golang/convert"
+	"github.com/spdx/tools-golang/spdx/common"
+	v2common "github.com/spdx/tools-golang/spdx/v2/common"
+	"github.com/spdx/tools-golang/spdx/v2/v2_1"
+	"github.com/spdx/tools-golang/spdx/v2/v2_2"
+	"github.com/spdx/tools-golang/spdx/v2/v2_3"
+)
+
+type Version string
+
+const (
+	V2_1 Version = "v2.1"
+	V2_2 Version = "v2.2"
+	V2_3 Version = "v2.3"
+)
+
+type Format string
+
+const JSON Format = "json"
+
+type Creator struct {
+	Creator string
+	// In accordance to the SPDX v2 spec, CreatorType should be one of "Person", "Organization", or "Tool"
+	CreatorType string
+}
+
+var _ sbom.Encoder = (*Encoder)(nil)
+
+type Encoder struct {
+	Version           Version
+	Format            Format
+	Creators          []Creator
+	DocumentNamespace string
+	DocumentComment   string
+}
+
+// Encode encodes a claircore IndexReport to an io.Reader.
+// We first convert the IndexReport to an SPDX doc of the latest version, then
+// convert that doc to the specified version. We assume there's no data munging
+// going from latest to the specified version.
+func (e *Encoder) Encode(ctx context.Context, ir *claircore.IndexReport) (io.Reader, error) {
+	spdx, err := e.parseIndexReport(ctx, ir)
+	if err != nil {
+		return nil, err
+	}
+
+	var tmpConverterDoc common.AnyDocument
+	switch e.Version {
+	case V2_1:
+		var targetDoc v2_1.Document
+		if err := convert.Document(spdx, targetDoc); err != nil {
+			return nil, err
+		}
+		tmpConverterDoc = targetDoc
+	case V2_2:
+		var targetDoc v2_2.Document
+		if err := convert.Document(spdx, targetDoc); err != nil {
+			return nil, err
+		}
+		tmpConverterDoc = targetDoc
+	case V2_3:
+		// parseIndexReport currently returns a v2_3.Document so do nothing
+		tmpConverterDoc = spdx
+	default:
+		return nil, fmt.Errorf("unknown SPDX version: %v", e.Version)
+	}
+
+	switch e.Format {
+	case JSON:
+		// TODO(DO NOT MERGE): Should this be outside the switch? i.e., should
+		//  we use this for all formatting cases? We could return it outside the
+		//  switch as the "default" case
+		buf := &bytes.Buffer{}
+		if err := spdxjson.Write(tmpConverterDoc, buf); err != nil {
+			return nil, err
+		}
```

Comment:
With the error handling IMHO it's OK either way, inside the switch or outside. It does seem the most common format is `json`, making that the 'default' could make consumption a tiny bit easier.